### ISSUE TITLE
Builder cannot connect to internal docker registry

### ISSUE THREAD

**Issue by kylegao91** [polyaxon/polyaxon#202]
### Describe the bug
The builder logs the following when trying to run an experiment.  Looks like the local docker daemon is not running?
```
ERROR 2018-10-10 15:49:43,783 builder 1 139804275205888 Failed to connect to registry 500 Server Error: Internal Server Error ("Get http://127.0.0.1:31813/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)")
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/docker/api/client.py", line 229, in _raise_for_status
    response.raise_for_status()
  File "/usr/local/lib/python3.5/dist-packages/requests/models.py", line 939, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http+docker://localhost/v1.38/auth

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/polyaxon/polyaxon/dockerizer/builder.py", line 79, in login_internal_registry
    reauth=True)
  File "/usr/local/lib/python3.5/dist-packages/docker/api/daemon.py", line 152, in login
    return self._result(response, json=True)
  File "/usr/local/lib/python3.5/dist-packages/docker/api/client.py", line 235, in _result
    self._raise_for_status(response)
  File "/usr/local/lib/python3.5/dist-packages/docker/api/client.py", line 231, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/usr/local/lib/python3.5/dist-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.APIError: 500 Server Error: Internal Server Error ("Get http://127.0.0.1:31813/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)")
```

### To Reproduce
Haven't found a good way to reproduce.  This is on an on-premise k8s cluster instead of minikube.  Actually we have 2 clusters and polyaxon has this problem only on one of the two.  I can dig in the difference between two clusters to provide more info.  

### Expected behavior
Successfully build the docker and launch the experiment.

### Environment
K8s cluster: 
```
Cluster info:

--------------  ----------------------------------------
minor           12
platform        linux/amd64
git_tree_state  clean
compiler        gc
major           1
git_commit      0ed33881dc4355495f623c6f22e7dd0b7632b7c0
build_date      2018-09-27T16:55:41Z
git_version     v1.12.0
go_version      go1.10.4
--------------  ----------------------------------------
```

Polyaxon: 
  server chart: polyaxon-0.2.6
  cli: 0.2.7

config:
```
namespace: polyaxon
rbac:
  enabled: false
ingress:
  enabled: false
serviceType: NodePort
limitResources: true
persistence:
  logs: {}
  repos: {}
  upload: {}
  data: {}
  outputs: {}
mountPaths:
  nvidia:
    lib: /usr/local/nvidia/lib
    bin: /usr/local/nvidia/bin
    libcuda: /usr/local/nvidia/lib64
dirs:
  nvidia:
    lib: xxx
    bin: yyy
    libcuda: zzz
```

**Comment by mouradmourafiq** [comment]
@kylegao91 Do you run both deployments with the same config and resources? can you check as well that the registry is running correctly in the failing cluster?

**Comment by kylegao91** [comment]
Yes the two deployments use the same config.
I checked that the registry is running.  Actually the 5000 port is working when I curl the cluster IP of the registry, but the nodeport port 31813 doesn't work.  Could it be the kube-proxy that's not working?

**Comment by mouradmourafiq** [comment]
Yes, it could be an issue with networking. On the cluster where Polyaxon is working correctly, can you check if the build pods are scheduled on the same node as the registry or on a different node.

**Comment by kylegao91** [comment]
No, the build pods and the registry pod are on different node in both clusters.  My confusion is that why it is trying to reach `127.0.0.1` instead of the IP of the node it's on?

**Comment by mouradmourafiq** [comment]
It's because docker does not raise for localhost ip with http, but would raise for any non localhost ip if not used with https. So that usage is intended, in some deployment, it will hang.

**Comment by mouradmourafiq** [comment]
Did you build your k8s clusters the same way? are they using same networking?

**Comment by kylegao91** [comment]
I don't have control of the cluster builds, but I will ask and get that info. 

On the other hand, I'm able to to reach the registry by curling the IP of the node where the build job runs on: `curl node-ip:31813/v2/_catalog`
How is the host port forwarded to the build job containers?

**Comment by kylegao91** [comment]
I found that the network is actually working by trying this:
```
curl -v   --unix-socket /var/run/docker.sock   "http://localhost/v1.38/auth"   -H "Content-Type: application/json"   -d '{"password":"", "username":""}'
*   Trying /var/run/docker.sock...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /v1.38/auth HTTP/1.1
> Host: localhost
> User-Agent: curl/7.47.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 75
>
* upload completely sent off: 75 out of 75 bytes
< HTTP/1.1 200 OK
< Api-Version: 1.38
< Content-Type: application/json
< Docker-Experimental: false
< Ostype: linux
< Server: Docker/18.06.1-ce (linux)
< Date: Mon, 15 Oct 2018 14:18:25 GMT
< Content-Length: 48
<
{"IdentityToken":"","Status":"Login Succeeded"}
* Connection #0 to host localhost left intact
```

**Comment by clark-wu** [comment]
Have you solved this problem? I have the same problem

**Comment by eublefar** [comment]
I have the same problem too

**Comment by zewelor** [comment]
Admin in our company said this solved this issue for us: https://github.com/kubernetes/kubernetes/issues/67730

**Comment by VinayTeki** [comment]
I'm trying solve a similar issues. 
Issue description:

Deployment of polyaxon on microk8s

I've deployed polyaxon on microk8s following these steps:


```bash

# the polyaxon deployment script

microk8s.kubectl create namespace polyaxon
sudo helm repo add polyaxon https://charts.polyaxon.com --kubeconfig kubeconfig_local.yml
sudo helm repo update --kubeconfig kubeconfig_local.yml
sudo helm install polyaxon/polyaxon \
    --name=polyaxon \
    --namespace=polyaxon -f poly-config.yaml \
    --kubeconfig kubeconfig_local.yml

#  what is  kubeconfig_local.yaml ? this is explained below
# Also what is in the file poly-config.yaml ? this is also explained below

```

```yaml
# File: kubeconfig_local.yml
apiVersion: v1
clusters:
- cluster:
    server: http://127.0.0.1:8080
  name: microk8s-cluster
contexts:
- context:
    cluster: microk8s-cluster
    user: admin
  name: microk8s
current-context: microk8s
kind: Config
preferences: {}
users:
- name: admin
  user:
    username: admin
```

```yaml
# File: poly-config.yml 
serviceType: NodePort

persistence:
  logs:
    existingClaim:
    mountPath: "/polyaxon-logs"
    hostPath: "/tmp/polyaxon/logs"
  repos:
    existingClaim:
    mountPath: "/polyaxon-repos"
    hostPath: "/tmp/polyaxon/repos"
  upload:
    existingClaim:
    mountPath: "/tmp/upload"
    hostPath: "/tmp/polyaxon/upload"
  data:
    data1:
      mountPath: "/data/1"
      hostPath: "/tmp/polyaxon/data"
      readOnly: false
  outputs:
    outputs1:
      mountPath: "/outputs/1"
      hostPath: "/tmp/polyaxon/outputs"
      readOnly: false

```

The deployment works fine with no issues at all

But when I run an experiment, container image gets built with no issues but cannot start the experiment.

```bash
NAME                                                READY   STATUS             RESTARTS   AGE
pod/plx-0d296e3a3fca4ff487c89b08471efe6c-master-0   1/2     ImagePullBackOff   0          36m
pod/polyaxon-docker-registry-7f7f7bb68-v674z        1/1     Running            0          42m
pod/polyaxon-polyaxon-api-7f9944bb6f-zkq9c          2/2     Running            0          42m

```

I've pulled the pod description using the command 

```bash
microk8s.kubectl describe pods plx-0d296e3a3fca4ff487c89b08471efe6c-master-0 -n polyaxon

# The output of this command is this (Pasting only the EVENTS part of the logs):

Events:
  Type     Reason     Age                   From                  Message
  ----     ------     ----                  ----                  -------
  Normal   Scheduled  41m                   default-scheduler     Successfully assigned polyaxon/plx-0d296e3a3fca4ff487c89b08471efe6c-master-0 to thingxthing
  Normal   Pulling    41m                   kubelet, thingxthing  Pulling image "polyaxon/polyaxon-init:0.4.4"
  Normal   Pulled     41m                   kubelet, thingxthing  Successfully pulled image "polyaxon/polyaxon-init:0.4.4"
  Normal   Created    41m                   kubelet, thingxthing  Created container polyaxon-job-init
  Normal   Started    41m                   kubelet, thingxthing  Started container polyaxon-job-init
  Normal   Pulling    41m                   kubelet, thingxthing  Pulling image "polyaxon/polyaxon-sidecar:0.4.4"
  Normal   Pulled     41m                   kubelet, thingxthing  Successfully pulled image "polyaxon/polyaxon-sidecar:0.4.4"
  Normal   Started    41m                   kubelet, thingxthing  Started container polyaxon-job-sidecar
  Normal   Created    41m                   kubelet, thingxthing  Created container polyaxon-job-sidecar
  Warning  Failed     40m (x3 over 41m)     kubelet, thingxthing  Error: ErrImagePull
  Warning  Failed     40m (x3 over 41m)     kubelet, thingxthing  Failed to pull image "127.0.0.1:31813/quick_1:d1741febeec94d5190c5ea3f65b1e69f": rpc error: code = Unknown desc = failed to resolve image "127.0.0.1:31813/quick_1:d1741febeec94d5190c5ea3f65b1e69f": no available registry endpoint: failed to do request: Head https://127.0.0.1:31813/v2/quick_1/manifests/d1741febeec94d5190c5ea3f65b1e69f: http: server gave HTTP response to HTTPS client
  Normal   Pulling    40m (x3 over 41m)     kubelet, thingxthing  Pulling image "127.0.0.1:31813/quick_1:d1741febeec94d5190c5ea3f65b1e69f"
  Warning  Failed     40m (x4 over 41m)     kubelet, thingxthing  Error: ImagePullBackOff
  Normal   BackOff    103s (x175 over 41m)  kubelet, thingxthing  Back-off pulling image "127.0.0.1:31813/quick_1:d1741febeec94d5190c5ea3f65b1e69f"

```

Now in the above logs this is the part I am concerned about 
```bash
no available registry endpoint: failed to do request: Head https://127.0.0.1:31813/v2/quick_1/manifests/d1741febeec94d5190c5ea3f65b1e69f: http: server gave HTTP response to HTTPS client
```

The error message says that it is trying to request the docker-registry 
endpoint with a https head where as it only accepts http head

So I tried to check if that is the case. I'm not sure if this means anything but I feel its worth taking a look at.

```bash
 >> curl -i http://127.0.0.1:31813 # curling http endpoint

    HTTP/1.1 200 OK
    Cache-Control: no-cache
    Date: Sun, 14 Apr 2019 10:03:20 GMT
    Content-Length: 0
    Content-Type: text/plain; charset=utf-8

 >> curl -i https://127.0.0.1:31813 # curling https endpoint
    
    curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number


```


I assumed docker should be informed to handle these connections and I've added this line to

```bash
# In file: /etc/default/docker

# did not know which one to use, so I tried all of the below
DOCKER_OPTS='--insecure-registry 127.0.0.1:31813' # did not work


# Here the ip is cluster ip of docker-registry service of polyaxon, didn't work either
NAME                               TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                 AGE
service/polyaxon-docker-registry   NodePort    10.152.183.37    <none>        5000:31813/TCP                          94m

DOCKER_OPTS='--insecure-registry 10.152.183.37:31813' 

# After adding this i restarted the docker service on my machine
>> sudo service docker restart

# I was expecting this should solve the problem but didn't
```

Help is very much appreciated, I've already spent a couple of days trying to solve this but couldn't find any solution or atleast a work around.

Thanks in advance.

**Comment by mouradmourafiq** [comment]
There seem to be an [issue](https://github.com/ubuntu/microk8s/issues/384) with new version of microk8s, I know that someone already managed to find a workaround, and another person is in the process to write a guide that will be added to the documentation, see #424

**Comment by VinayTeki** [comment]
@mouradmourafiq  thanks for the suggestion will try the ideas presented in that issue.

**Comment by sotte** [comment]
I'll wait till this issues is solved (no kubexpert here :)) and then update #424. 
@VinayTeki I was not able to solve my problem based on https://github.com/ubuntu/microk8s/issues/384 but @maltebaumann was I think.

**Comment by mouradmourafiq** [comment]
Closing this issue, feel free to reopen.
